
project system/sepolicy/
diff --git a/build/soong/service_fuzzer_bindings.go b/build/soong/service_fuzzer_bindings.go
index 15e885a..da71147 100644
--- a/build/soong/service_fuzzer_bindings.go
+++ b/build/soong/service_fuzzer_bindings.go
@@ -23,6 +23,7 @@ var EXCEPTION_NO_FUZZER = []string{}
 
 var (
 	ServiceFuzzerBindings = map[string][]string{
+		"adbroot_service":                                                                   EXCEPTION_NO_FUZZER,
 		"android.hardware.audio.core.IConfig/default":                                        EXCEPTION_NO_FUZZER,
 		"android.hardware.audio.core.IModule/default":                                        EXCEPTION_NO_FUZZER,
 		"android.hardware.audio.core.IModule/a2dp":                                           EXCEPTION_NO_FUZZER,
diff --git a/contexts/file_contexts_test_data b/contexts/file_contexts_test_data
index 06a8ca1..f2e971d 100644
--- a/contexts/file_contexts_test_data
+++ b/contexts/file_contexts_test_data
@@ -793,6 +793,8 @@
 /system_ext/bin/hw/android.hidl.allocator@1.0-service             hal_allocator_default_exec
 /system/system_ext/bin/hw/android.hidl.allocator@1.0-service      hal_allocator_default_exec
 
+/system_ext/bin/adb_root                                          adbroot_exec
+/system/system_ext/bin/adb_root                                   adbroot_exec
 
 /system_ext/bin/rkp_cert_processor                                rkp_cert_processor_exec
 /system/system_ext/bin/rkp_cert_processor                         rkp_cert_processor_exec
@@ -878,6 +880,8 @@
 /data/ota_package/test                                            ota_package_file
 /data/adb                                                         adb_data_file
 /data/adb/test                                                    adb_data_file
+/data/adbroot                                                     adbroot_data_file
+/data/adbroot/test                                                adbroot_data_file
 /data/anr                                                         anr_data_file
 /data/anr/test                                                    anr_data_file
 /data/apex                                                        apex_data_file
diff --git a/prebuilts/api/202504/private/adbd.te b/prebuilts/api/202504/private/adbd.te
index 138c8f5..5f8ef67 100644
--- a/prebuilts/api/202504/private/adbd.te
+++ b/prebuilts/api/202504/private/adbd.te
@@ -14,6 +14,8 @@ allow adbd adbd_tradeinmode:process dyntransition;
 
 userdebug_or_eng(`
   allow adbd su:process dyntransition;
+  allow adbd adbroot:binder call;
+  allow adbd adbroot_service:service_manager find;
 ')
 
 # When 'adb shell' is executed in recovery mode, adbd explicitly
diff --git a/prebuilts/api/202504/private/adbroot.te b/prebuilts/api/202504/private/adbroot.te
new file mode 100644
index 0000000..828ce90
--- /dev/null
+++ b/prebuilts/api/202504/private/adbroot.te
@@ -0,0 +1,17 @@
+type adbroot, domain, coredomain;
+type adbroot_exec, exec_type, file_type, system_file_type;
+
+init_daemon_domain(adbroot)
+
+binder_use(adbroot)
+binder_service(adbroot)
+add_service(adbroot, adbroot_service)
+
+allow adbroot package_native_service:service_manager find;
+binder_call(adbroot, system_server)
+
+allow adbroot adbroot_data_file:dir rw_dir_perms;
+allow adbroot adbroot_data_file:file create_file_perms;
+
+set_prop(adbroot, shell_prop)
+set_prop(adbroot, ctl_adbd_prop)
diff --git a/prebuilts/api/202504/private/file.te b/prebuilts/api/202504/private/file.te
index 6bdcc39..e44c073 100644
--- a/prebuilts/api/202504/private/file.te
+++ b/prebuilts/api/202504/private/file.te
@@ -14,6 +14,9 @@ type fs_bpf_loader, fs_type, bpffs_type;
 type fs_bpf_uprobestats, fs_type, bpffs_type;
 type fs_bpf_memevents, fs_type, bpffs_type;
 
+# /data/adbroot
+type adbroot_data_file, file_type, data_file_type, core_data_file_type;
+
 # /data/misc/storaged
 type storaged_data_file, file_type, data_file_type, core_data_file_type;
 
diff --git a/prebuilts/api/202504/private/file_contexts b/prebuilts/api/202504/private/file_contexts
index 0b3e7f4..4d8e8fa 100644
--- a/prebuilts/api/202504/private/file_contexts
+++ b/prebuilts/api/202504/private/file_contexts
@@ -543,6 +543,7 @@
 
 /(system_ext|system/system_ext)/bin/aidl_lazy_test_server    u:object_r:aidl_lazy_test_server_exec:s0
 /(system_ext|system/system_ext)/bin/aidl_lazy_cb_test_server u:object_r:aidl_lazy_test_server_exec:s0
+/(system_ext|system/system_ext)/bin/adb_root                u:object_r:adbroot_exec:s0
 /(system_ext|system/system_ext)/bin/hidl_lazy_test_server    u:object_r:hidl_lazy_test_server_exec:s0
 /(system_ext|system/system_ext)/bin/hidl_lazy_cb_test_server u:object_r:hidl_lazy_test_server_exec:s0
 /(system_ext|system/system_ext)/bin/hwservicemanager         u:object_r:hwservicemanager_exec:s0
@@ -599,6 +600,7 @@
 /data/dalvik-cache(/.*)? u:object_r:dalvikcache_data_file:s0
 /data/ota(/.*)? u:object_r:ota_data_file:s0
 /data/ota_package(/.*)? u:object_r:ota_package_file:s0
+/data/adbroot(/.*)?	u:object_r:adbroot_data_file:s0
 /data/adb(/.*)?		u:object_r:adb_data_file:s0
 /data/anr(/.*)?		u:object_r:anr_data_file:s0
 /data/apex(/.*)?		u:object_r:apex_data_file:s0
diff --git a/prebuilts/api/202504/private/platform_app.te b/prebuilts/api/202504/private/platform_app.te
index 38339da..0a7a5f4 100644
--- a/prebuilts/api/202504/private/platform_app.te
+++ b/prebuilts/api/202504/private/platform_app.te
@@ -100,6 +100,7 @@ allow platform_app system_api_service:service_manager find;
 allow platform_app vr_manager_service:service_manager find;
 allow platform_app stats_service:service_manager find;
 allow platform_app tradeinmode_service:service_manager find;
+userdebug_or_eng(`allow platform_app adbroot_service:service_manager find;')
 
 # Allow platform apps to log via statsd.
 binder_call(platform_app, statsd)
diff --git a/prebuilts/api/202504/private/service.te b/prebuilts/api/202504/private/service.te
index c17f49e..c7e0647 100644
--- a/prebuilts/api/202504/private/service.te
+++ b/prebuilts/api/202504/private/service.te
@@ -6,6 +6,7 @@ until_board_api(202504, `
     type app_function_service, app_api_service, system_server_service, service_manager_type;
 ')
 type attention_service,              system_server_service, service_manager_type;
+type adbroot_service,                service_manager_type;
 type bg_install_control_service,     system_api_service, system_server_service, service_manager_type;
 type compos_service,                 service_manager_type;
 type communal_service,               app_api_service, system_server_service, service_manager_type;
diff --git a/prebuilts/api/202504/private/service_contexts b/prebuilts/api/202504/private/service_contexts
index 9edea85..6f1da47 100644
--- a/prebuilts/api/202504/private/service_contexts
+++ b/prebuilts/api/202504/private/service_contexts
@@ -151,6 +151,7 @@ accessibility                             u:object_r:accessibility_service:s0
 account                                   u:object_r:account_service:s0
 activity                                  u:object_r:activity_service:s0
 activity_task                             u:object_r:activity_task_service:s0
+adbroot_service                           u:object_r:adbroot_service:s0
 MiuiInit                                  u:object_r:miui_init_service:s0
 adb                                       u:object_r:adb_service:s0
 adservices_manager                        u:object_r:adservices_manager_service:s0
diff --git a/prebuilts/api/202504/private/system_server.te b/prebuilts/api/202504/private/system_server.te
index 4e5112e..0ffe0ec 100644
--- a/prebuilts/api/202504/private/system_server.te
+++ b/prebuilts/api/202504/private/system_server.te
@@ -10,6 +10,7 @@ typeattribute system_server scheduler_service_server;
 typeattribute system_server sensor_service_server;
 typeattribute system_server stats_service_server;
 typeattribute system_server bpfdomain;
+userdebug_or_eng(`allow system_server adbroot_service:service_manager find;')
 
 # Define a type for tmpfs-backed ashmem regions.
 tmpfs_domain(system_server)
diff --git a/private/adbd.te b/private/adbd.te
index 6bef6f8..f0c15af 100644
--- a/private/adbd.te
+++ b/private/adbd.te
@@ -17,6 +17,8 @@ allow adbd adbd_tradeinmode:process dyntransition;
 
 userdebug_or_eng(`
   allow adbd su:process dyntransition;
+  allow adbd adbroot:binder call;
+  allow adbd adbroot_service:service_manager find;
 ')
 
 # When 'adb shell' is executed in recovery mode, adbd explicitly
diff --git a/private/adbroot.te b/private/adbroot.te
new file mode 100644
index 0000000..828ce90
--- /dev/null
+++ b/private/adbroot.te
@@ -0,0 +1,17 @@
+type adbroot, domain, coredomain;
+type adbroot_exec, exec_type, file_type, system_file_type;
+
+init_daemon_domain(adbroot)
+
+binder_use(adbroot)
+binder_service(adbroot)
+add_service(adbroot, adbroot_service)
+
+allow adbroot package_native_service:service_manager find;
+binder_call(adbroot, system_server)
+
+allow adbroot adbroot_data_file:dir rw_dir_perms;
+allow adbroot adbroot_data_file:file create_file_perms;
+
+set_prop(adbroot, shell_prop)
+set_prop(adbroot, ctl_adbd_prop)
diff --git a/private/file.te b/private/file.te
index b832da4..6da616a 100644
--- a/private/file.te
+++ b/private/file.te
@@ -21,6 +21,9 @@ type fs_bpf_cputimeinstate, fs_type, bpffs_type;
 # /data/system/mediadrm
 type mediadrm_system_data_file, file_type, data_file_type, core_data_file_type;
 
+# /data/adbroot
+type adbroot_data_file, file_type, data_file_type, core_data_file_type;
+
 # /data/misc/storaged
 type storaged_data_file, file_type, data_file_type, core_data_file_type;
 
diff --git a/private/file_contexts b/private/file_contexts
index bc79eeb..a82ea18 100644
--- a/private/file_contexts
+++ b/private/file_contexts
@@ -545,6 +545,7 @@
 
 /(system_ext|system/system_ext)/bin/aidl_lazy_test_server    u:object_r:aidl_lazy_test_server_exec:s0
 /(system_ext|system/system_ext)/bin/aidl_lazy_cb_test_server u:object_r:aidl_lazy_test_server_exec:s0
+/(system_ext|system/system_ext)/bin/adb_root                u:object_r:adbroot_exec:s0
 /(system_ext|system/system_ext)/bin/hidl_lazy_test_server    u:object_r:hidl_lazy_test_server_exec:s0
 /(system_ext|system/system_ext)/bin/hidl_lazy_cb_test_server u:object_r:hidl_lazy_test_server_exec:s0
 /(system_ext|system/system_ext)/bin/hwservicemanager         u:object_r:hwservicemanager_exec:s0
@@ -605,6 +606,7 @@
 /data/resource-cache(/.*)? u:object_r:resourcecache_data_file:s0
 /data/dalvik-cache(/.*)? u:object_r:dalvikcache_data_file:s0
 /data/ota_package(/.*)? u:object_r:ota_package_file:s0
+/data/adbroot(/.*)?	u:object_r:adbroot_data_file:s0
 /data/adb(/.*)?		u:object_r:adb_data_file:s0
 /data/anr(/.*)?		u:object_r:anr_data_file:s0
 /data/apex(/.*)?		u:object_r:apex_data_file:s0
diff --git a/private/platform_app.te b/private/platform_app.te
index 9e859f4..ea5e19d 100644
--- a/private/platform_app.te
+++ b/private/platform_app.te
@@ -101,6 +101,7 @@ allow platform_app system_api_service:service_manager find;
 allow platform_app vr_manager_service:service_manager find;
 allow platform_app stats_service:service_manager find;
 allow platform_app tradeinmode_service:service_manager find;
+userdebug_or_eng(`allow platform_app adbroot_service:service_manager find;')
 
 # Allow platform apps to log via statsd.
 binder_call(platform_app, statsd)
diff --git a/private/service.te b/private/service.te
index 331fbac..62e2e70 100644
--- a/private/service.te
+++ b/private/service.te
@@ -3,6 +3,7 @@ type ambient_context_service,        app_api_service, system_server_service, ser
 type authentication_policy_service,  app_api_service, system_api_service, system_server_service, service_manager_type;
 
 type attention_service,              system_server_service, service_manager_type;
+type adbroot_service,                service_manager_type;
 type bg_install_control_service,     system_api_service, system_server_service, service_manager_type;
 type binder_stats_consumer_service,  system_server_service, service_manager_type;
 type compos_service,                 service_manager_type;
diff --git a/private/service_contexts b/private/service_contexts
index 361c89a..6e61a70 100644
--- a/private/service_contexts
+++ b/private/service_contexts
@@ -157,6 +157,7 @@ account                                   u:object_r:account_service:s0
 activity                                  u:object_r:activity_service:s0
 activity_structured                       u:object_r:activity_structured_service:s0
 activity_task                             u:object_r:activity_task_service:s0
+adbroot_service                           u:object_r:adbroot_service:s0
 MiuiInit                                  u:object_r:miui_init_service:s0
 adb                                       u:object_r:adb_service:s0
 adservices_manager                        u:object_r:adservices_manager_service:s0
diff --git a/private/system_server.te b/private/system_server.te
index d85caed..229fdc7 100644
--- a/private/system_server.te
+++ b/private/system_server.te
@@ -10,6 +10,7 @@ typeattribute system_server scheduler_service_server;
 typeattribute system_server sensor_service_server;
 typeattribute system_server stats_service_server;
 typeattribute system_server bpfdomain;
+userdebug_or_eng(`allow system_server adbroot_service:service_manager find;')
 
 # Define a type for tmpfs-backed ashmem regions.
 tmpfs_domain(system_server)

project device/lineage/sepolicy/
diff --git a/common/private/adbd.te b/common/private/adbd.te
deleted file mode 100644
index e5ef587..0000000
--- a/common/private/adbd.te
+++ /dev/null
@@ -1,2 +0,0 @@
-allow adbd adbroot:binder call;
-allow adbd adbroot_service:service_manager find;
diff --git a/common/private/adbroot.te b/common/private/adbroot.te
deleted file mode 100644
index 828ce90..0000000
--- a/common/private/adbroot.te
+++ /dev/null
@@ -1,17 +0,0 @@
-type adbroot, domain, coredomain;
-type adbroot_exec, exec_type, file_type, system_file_type;
-
-init_daemon_domain(adbroot)
-
-binder_use(adbroot)
-binder_service(adbroot)
-add_service(adbroot, adbroot_service)
-
-allow adbroot package_native_service:service_manager find;
-binder_call(adbroot, system_server)
-
-allow adbroot adbroot_data_file:dir rw_dir_perms;
-allow adbroot adbroot_data_file:file create_file_perms;
-
-set_prop(adbroot, shell_prop)
-set_prop(adbroot, ctl_adbd_prop)
diff --git a/common/private/file.te b/common/private/file.te
index b57ab7f..d359cef 100644
--- a/common/private/file.te
+++ b/common/private/file.te
@@ -1,5 +1,4 @@
 type sdcard_posix, sdcard_type, sdcard_posix_contextmount_type, fs_type, mlstrustedobject;
-type adbroot_data_file, file_type, data_file_type, core_data_file_type;
 type sysfs_perdev_minors, fs_type, sysfs_type;
 type lineage_platform_framework_file, file_type, system_file_type;
 type lineage_permission_file, file_type, system_file_type;
diff --git a/common/private/file_contexts b/common/private/file_contexts
index cbf1a53..9fda3a2 100644
--- a/common/private/file_contexts
+++ b/common/private/file_contexts
@@ -58,9 +58,5 @@
 /system/bin/backuptool_ab\.sh                     u:object_r:postinstall_exec:s0
 /system/bin/backuptool_postinstall\.sh            u:object_r:postinstall_exec:s0
 
-# ADB Root
-/(system_ext|system/system_ext)/bin/adb_root      u:object_r:adbroot_exec:s0
-/data/adbroot(/.*)?        u:object_r:adbroot_data_file:s0
-
 # Bash
 /(system_ext|system/system_ext)/bin/bash          u:object_r:shell_exec:s0
diff --git a/common/private/service.te b/common/private/service.te
index 86f6e4c..2d32e4c 100644
--- a/common/private/service.te
+++ b/common/private/service.te
@@ -1,4 +1,3 @@
-type adbroot_service,                   service_manager_type;
 type lineage_hardware_service, system_api_service, system_server_service, service_manager_type;
 type lineage_health_interface_service, system_api_service, system_server_service, service_manager_type;
 type lineage_globalactions_service, system_api_service, system_server_service, service_manager_type;
diff --git a/common/private/service_contexts b/common/private/service_contexts
index 0a6a8b0..16d437b 100644
--- a/common/private/service_contexts
+++ b/common/private/service_contexts
@@ -4,4 +4,3 @@ healthinterface                           u:object_r:lineage_health_interface_se
 displayservice                            u:object_r:lineage_livedisplay_service:s0
 trustinterface                            u:object_r:lineage_trust_service:s0
 profile                                   u:object_r:lineage_profile_service:s0
-adbroot_service                           u:object_r:adbroot_service:s0
diff --git a/common/private/system_server.te b/common/private/system_server.te
index 8fd4417..9768f36 100644
--- a/common/private/system_server.te
+++ b/common/private/system_server.te
@@ -1,5 +1,4 @@
 allow system_server storage_stub_file:dir getattr;
-allow system_server adbroot_service:service_manager find;
 
 # Allow SystemConfig/permission parsing to read relabeled Lineage permission XMLs.
 allow system_server lineage_permission_file:file r_file_perms;
